### ISSUE TITLE
Automatic cancellation support

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#904]
I have a lot of dependencies with (usually `async`) methods that accept a `CancellationToken`. When the token is cancelled, these methods throw an `OperationCanceledException`. This is the expected implicit behavior for these methods, but by default, fakes don't throw anything when they receive a call with a cancelled token. So my thinking is that this behavior could be baked into FakeItEasy as a default rule. Currently I use a method like this:

```c#
private static void ConfigureCancellation(object obj)
{
    A.CallTo(obj)
        .WhenArgumentsMatch(args => args.OfType<CancellationToken>().Any(ct => ct.IsCancellationRequested))
        .Invokes(call => call.Arguments.OfType<CancellationToken>().FirstOrDefault(ct => ct.IsCancellationRequested).ThrowIfCancellationRequested());
}
```

It works fine, but I have to remember to call it on each fake.

Note that enabling this by default would, technically, be a breaking change, although probably not one that would affect many users.

**Comment by blairconrad** [comment]
> It works fine, but I have to remember to call it on each fake.

I'm not trying to quash this issue, but have you tried to build the behaviour into a [FakeOptionsBuilder](http://fakeiteasy.readthedocs.io/en/stable/implicit-creation-options/)?

**Comment by blairconrad** [comment]
Also, while I'm critiquing, you might want `First` instead of `FirstOrDefault`, since there's going to be an item in the enumerable, and if there weren't you'd throw a less-helpful exception anyhow.

**Comment by thomaslevesque** [comment]
> I'm not trying to quash this issue, but have you tried to build the behaviour into a `FakeOptionsBuilder`?

It wouldn't really solve anything... I want *all* fakes to have this behavior, and I don't want to create a fake options builder for each type of dependency.

By convention (with very few exceptions), a method that accepts a `CancellationToken` should always throw if the token is cancelled. What I'd like is to implicitly enforce that convention in FakeItEasy.

> Also, while I'm critiquing, you might want `First` instead of `FirstOrDefault`, since there's going to be an item in the enumerable, and if there weren't you'd throw a less-helpful exception anyhow.

Good point!

**Comment by blairconrad** [comment]
> I want all fakes to have this behavior, and I don't want to create a fake options builder for each type of dependency.

I was thinking of something like

```c#
public class ThrowsWhenCancelledFakeOptionsBuilder: IFakeOptionsBuilder
{
    public bool CanBuildOptionsForFakeOfType(Type type) => true;

    public Priority Priority => Priority.Default;

    public void BuildOptions(Type typeOfFake, IFakeOptions options)
    {
        options.ConfigureFake(obj => A.CallTo(obj)
            .WhenArgumentsMatch(args =>
                args.OfType<CancellationToken>()
                    .Any(ct => ct.IsCancellationRequested))
            .Invokes(call =>
                call.Arguments.OfType<CancellationToken>()
                    .First(ct => ct.IsCancellationRequested)
                    .ThrowIfCancellationRequested()));
    }
}
```

Of course, you could get fancier in `CanBuildOptionsForFakeOfType` and look for methods that accept cancellation tokens. But I'm lazy.

**Comment by thomaslevesque** [comment]
Ah, I didn't think of using a non-generic fake options builder. But that wouldn't work either, because we only use one FOB for a given fake type. Any other FOB with a higher priority would preempt the default one.

(btw, it might be interesting to consider using multiple FOBs for a given fake type)

**Comment by blairconrad** [comment]
Ah, you're already using FOBs? That is a problem. You could in theory make a more complicated FOB that handles all your logic, but that sounds like it will become nightmarish.

I've off and on thought about the limitation of having only one active FOB per type, but I think allowing multiple to apply could be very confusing.

**Comment by thomaslevesque** [comment]
> Ah, you're already using FOBs?

Yes. But anyway, I realized my approach has annoying limitations: the generic "throw when token is cancelled" configuration doesn't magically combine with other call configurations. As soon as I configure a method explicitly, it overrides the previous rule, so it no longer throws when I call the method with a cancelled token... unless I constrain the token to not be cancelled in the new configuration.

Not sure if I'm making sense, so here's an example. If I do this:

```
ConfigureCancellation(fake);
A.CallTo(() => fake.MyMethod(A<CancellationToken>._)).Returns(42);
```

then it i will always return 42, even with a cancelled `CancellationToken`. I need to constrain the token:

```
ConfigureCancellation(fake);
A.CallTo(() => fake.MyMethod(A<CancellationToken>.That.Matches(ct => !ct.IsCancellationRequested))).Returns(42);
```

which is annoying...

> I think allowing multiple to apply could be very confusing.

Indeed

**Comment by blairconrad** [comment]
> the generic "throw when token is cancelled" configuration doesn't magically combine with other call configurations. As soon as I configure a method explicitly, it overrides the previous rule, so it no longer throws when I call the method with a cancelled token... unless I constrain the token to not be cancelled in the new configuration

Yeah, that was part of my concern about "nightmarish" FOB, or about combining FOBs.

Does this affect your desire for automated support for cancellation tokens?
If it were to happen, I'd guess that the best we could do was "unconfigured methods will throw when one of the arguments is a cancelled token", and as soon as you introduce your own configuration to a method, you'll have to account for the cancellation token.

**Comment by thomaslevesque** [comment]
> Does this affect your desire for automated support for cancellation tokens?

Well, ideally I'd still like to have it, but I'm not sure how it would work.

> If it were to happen, I'd guess that the best we could do was "unconfigured methods will throw when one of the arguments is a cancelled token", and as soon as you introduce your own configuration to a method, you'll have to account for the cancellation token.

The way I eventually handled it is by creating a `NotCancelled` extension method on `IWhereConfiguration<T>` which adds a condition on the token, so I can do something like this:

```
ConfigureCancellation(fake);
A.CallTo(() => fake.MyMethod(A<CancellationToken>._)).NotCancelled().Returns(42);
```
The `.NotCancelled()` adds a check on the cancellation token, to avoid overriding the "throw when token is cancelled" rule. I suppose we could use such an approach in FakeItEasy itself if we wanted to support this feature, but it's not perfect.

IIRC it's not currently possible to have a rule that is always evaluated before all others (including user rules). If we had that, automatic cancellation would be pretty easy to implement, and user-defined rules wouldn't override the cancellation behavior.

**Comment by thomaslevesque** [comment]
> IIRC it's not currently possible to have a rule that is always evaluated before all others (including user rules)

A quick look at the code shows that it is actually possible:
https://github.com/FakeItEasy/FakeItEasy/blob/master/src/FakeItEasy/Core/FakeManager.cs#L86-L87

Currently we only add an `EventRule` to it.

**Comment by thomaslevesque** [comment]
Not sure yet if it's something we'll want to do, but I'm going to POC it to get a better idea.

**Comment by adamralph** [comment]
Given that #906 was abandoned, should we withdraw this?

**Comment by thomaslevesque** [comment]
Well, the approach I proposed in my PR was probably too risky, but I'm still hoping we can find a way to improve the experience around cancellation. If there's no interest, though, feel free to close this issue.

**Comment by adamralph** [comment]
I'm fine with leaving it open if we still want to explore other ways of doing it.

**Comment by thomaslevesque** [comment]
Thinking about this again... We could change the default behavior for unconfigured methods to throw (or return a cancelled task) when a cancelled token is passed, but *not* prevent this behavior from being overridden as I had initially suggested (@blairconrad eventually convinced me this was a crazy idea).

This would still be a breaking change, but I think it would have a very low impact. In unit tests, unless you're specifically testing cancellation, you almost *never* use a cancelled token, instead you typically use `CancellationToken.None` or `default(CancellationToken)`. This scenario would not be affected. On the other hand, if you *are* testing the cancellation behavior of your code, you typically already configure the fakes to throw when appropriate. This scenario wouldn't be broken either. So most existing tests wouldn't be affected by the change.

Now, for the other issue I had mentioned, the fact that it's too easy to unintentionally override the cancellation behavior can be mitigated by the addition of a "shortcut" extension method on `IArgumentConstraintManager<CancellationToken>`:

```
interface IFoo
{
    int Bar(CancellationToken cancellationToken);
}

var foo = A.Fake<IFoo>();

A.CallTo(() => foo.Bar(A<CancellationToken>.That.IsNotCancelled())).Returns(42);
```

Of course, you only need to use `IsNotCancelled()` if you care about keeping the cancellation behavior of the fake. Most tests wouldn't need it, so they wouldn't be affected.

Does this approach sound more reasonable?

**Comment by blairconrad** [comment]
I'm in. For all the things.

I see the value in having the default behaviour respect a cancellation token. My main fear was the change in semantics of explicitly configuring a method, and especially what it meant for `A<CancellationToken>._`. 

The only tweak I'd suggest (and you may already have thought of it, just not bothered mentioning it because it wasn't yet relevant) is to give `IsNotCancelled` a friend, `IsCancelled`.

**Comment by thomaslevesque** [comment]
Question: should async methods throw an `OperationCanceledException` directly, or return a cancelled task? From what I can see in the BCL (`Task.Delay`, `Stream.ReadAsync`, `SemaphoreSlim.WaitAsync`, `HttpClient.SendAsync`...), methods that receive an already cancelled token return a cancelled task, so I think we should do the same. But it makes the implementation a little more complex, because the tasks returned by the default behavior are actually dummies created by `FakeAndDummyManager`, and we have no mechanism to specify that we want a "dummy cancelled task" (we just specify the type of dummy that we want).

**Comment by blairconrad** [comment]
Returning a cancelled task sounds like the right thing to me.

I would've guessed that the new rule (and that's how I'd try to implement this, by adding a new post-user rule in the `FakeManager` constructor) would just return a newly-created cancelled task when the method is passed a cancelled token. (Or throw the exception if the return type is not a task.)

**Comment by adamralph** [comment]
👍 from me!

**Comment by thomaslevesque** [comment]
Since it is breaking, and we've already released a beta, is it too late to include it in 3.0.0?

**Comment by blairconrad** [comment]
NEVER!

**Comment by blairconrad** [comment]
More seriously, I certainly hope not. I'm not sure where the "beta == no more breaking changes" comes from, but it seems overly limiting to me. I do not know why we can't just keep breaking things during the pre-release phase.

Not that I'm advocating complete pandemonium. We'll be as responsible as we can, but I think it's silly to block progress just because we didn't anticipate this change when we were naming the previous release.

**Comment by thomaslevesque** [comment]
> I'm not sure where the "beta == no more breaking changes" comes from

@adamralph [mentioned it](https://github.com/FakeItEasy/FakeItEasy/issues/830#issuecomment-261801004) a few days ago, and it was a little surprising to me.

> just because we didn't anticipate this change when we were naming the previous release.

Actually, that's my fault. I might have been hasty in [saying](https://github.com/FakeItEasy/FakeItEasy/issues/919#issuecomment-263807345) this issue shouldn't be included in any release yet... I didn't think we'd be so quick to reach an agreement.

Of course, there's no rush for this feature; if you prefer to wait for 4.0.0, that's fine by me. If we want to include it in 3.0.0, I'll have a PR ready soon.

**Comment by blairconrad** [comment]
> @adamralph mentioned it a few days ago, and it was a little surprising to me.

Oh, sorry. I was unclear. I meant I didn't know the provenance. I was alluding to his comment.

> Actually, that's my fault. I might have been hasty in saying this issue shouldn't be included in any release yet... I didn't think we'd be so quick to reach an agreement.

Pfah! Don't be so hard on yourself.

> Of course, there's no rush for this feature; if you prefer to wait for 4.0.0, that's fine by me. If we want to include it in 3.0.0, I'll have a PR ready soon.

I'd much rather get it in now.

**Comment by adamralph** [comment]
I'm fine with it being part of 3.0.0.

There's no firm definition of the phases, but the way I see it is:

- alpha: anything may change before RTM and probably will.
- beta: things may still change before RTM but they probably won't.
- RC: RTM quality. *Will be the RTM if no bugs are found*. The only changes before RTM will be bug fixes and they will generate a new RC each time.

I think beta is the "grey zone" and is the tough one to call. Alpha and RC are relatively easy. I think it's reasonable to call beta if nothing more is planned to change *at that time the beta is released*, and this is what we did in this case.

**Comment by blairconrad** [comment]
>  I think it's reasonable to call beta if nothing more is planned to change _at the time the beta is released,_ and this is what we did in this case.

:+1:

**Comment by blairconrad** [comment]
This issue has been released in FakeItEasy 3.0.0:
https://github.com/FakeItEasy/FakeItEasy/releases/3.0.0
